solvers/minion/src/run.rs

```diff
@@ -1,29 +1,165 @@
-
-use crate::{
-    ast::*,
-    raw_bindings::{self, *},
-    scoped_ptr::Scoped,
+use std::{
+    collections::HashMap,
+    ffi::CString,
+    sync::{Mutex, MutexGuard},
 };
 
+use crate::{ast::*, error::*, raw_bindings::*, scoped_ptr::Scoped};
+
 // TODO: allow passing of options.
 
 /// Callback function used to capture results from minion as they are generated.
-/// Should return true if search is to continue, false otherwise.
-pub type Callback = fn(Vec<(VarName, Constant)>) -> bool;
+/// Should return `true` if search is to continue, `false` otherwise.
+///
+/// Consider using a global mutex (or other static variable) to use these results
+/// elsewhere.
+///
+/// For example:
+///
+/// ```
+///   use minion_rs::ast::*;
+///   use minion_rs::run_minion;
+///   use std::{
+///       collections::HashMap,
+///       sync::{Mutex, MutexGuard},
+///   };
+///
+///   // More elaborate data-structures are possible, but for sake of example store
+///   // a vector of solution sets.
+///   static ALL_SOLUTIONS: Mutex<Vec<HashMap<VarName,Constant>>>  = Mutex::new(vec![]);
+///   
+///   fn callback(solutions: HashMap<VarName,Constant>) -> bool {
+///       let mut guard = ALL_SOLUTIONS.lock().unwrap();
+///       guard.push(solutions);
+///       return true;
+///   }
+///    
+///   // Build and run the model.
+///   let mut model = Model::new();
+///
+///   // ... omitted for brevity ...
+/// #
+/// # model
+/// #     .named_variables
+/// #     .add_var("x".to_owned(), VarDomain::Bound(1, 3));
+/// # model
+/// #     .named_variables
+/// #     .add_var("y".to_owned(), VarDomain::Bound(2, 4));
+/// # model
+/// #     .named_variables
+/// #     .add_var("z".to_owned(), VarDomain::Bound(1, 5));
+/// #
+/// # let leq = Constraint::SumLeq(
+/// #     vec![
+/// #         Var::NameRef("x".to_owned()),
+/// #         Var::NameRef("y".to_owned()),
+/// #         Var::NameRef("z".to_owned()),
+/// #     ],
+/// #     Var::ConstantAsVar(4),
+/// # );
+/// #
+/// # let geq = Constraint::SumGeq(
+/// #     vec![
+/// #         Var::NameRef("x".to_owned()),
+/// #         Var::NameRef("y".to_owned()),
+/// #         Var::NameRef("z".to_owned()),
+/// #     ],
+/// #     Var::ConstantAsVar(4),
+/// # );
+/// #
+/// # let ineq = Constraint::Ineq(
+/// #     Var::NameRef("x".to_owned()),
+/// #     Var::NameRef("y".to_owned()),
+/// #     Constant::Integer(-1),
+/// # );
+/// #
+/// # model.constraints.push(leq);
+/// # model.constraints.push(geq);
+/// # model.constraints.push(ineq);
+///  
+///   let res = run_minion(model, callback);
+///   res.expect("Error occurred");
+///
+///   // Get solutions
+///   let guard = ALL_SOLUTIONS.lock().unwrap();
+///
+///   let solution_set_1 = &(*guard.get(0).unwrap());
+///
+///   let x1 = solution_set_1.get("x").unwrap();
+///   let y1 = solution_set_1.get("y").unwrap();
+///   let z1 = solution_set_1.get("z").unwrap();
+/// #
+/// # // FIXME: this test would be better with an example with >1 solution.
```

Comment:
just make one of the domains larger?

Reply:
I think I have Minion set to "one solution only". Also, as discussed on Teams, I am going to rethink how I am going to do examples generally